### ISSUE TITLE
oauth and 2FA error_description being returned in randomised languages

### ISSUE THREAD

**Issue by ligsnf** [thewca/worldcubeassociation.org#7735]
I'm new to using the WCA API and was testing the OAUTH using Postman. When entering invalid username or password in retrieving OAUTH token, the error_description would be returned in random languages on each request.

1. Go to POST https://www.worldcubeassociation.org/oauth/token
2. Enter valid client_id and client_secret
3. Enter invalid username or password
4. See error_description of response

Is the problem that a request location is not specified in the header?
Sorry if this has been explained before, I'm quite new to this
![Screenshot 2023-03-15 121830](https://user-images.githubusercontent.com/92285265/225179357-f40a8a76-7e30-491d-a98c-a31d42b769b3.png)
![Screenshot 2023-03-15 121817](https://user-images.githubusercontent.com/92285265/225179361-77485acf-3816-4bf9-8fe9-c97cc7852025.png)

**Comment by sarahstrong314** [comment]
FWIW, I've been experiencing the same issue.

**Comment by gregorbg** [comment]
FWIW, a similar issue happens for me with the 2FA code screen. It also always prompts my code in an utterly random language.

**Comment by dunkOnIT** [comment]
Are we using Devise for 2FA and OAuth? If so, perhaps there is a method we need to override?

**Comment by FinnIckler** [comment]
We use doorkeeper for providing oauth

**Comment by gregorbg** [comment]
Yeah, it's Devise (or rather, the devise-two-factor plugin to be precise) and Doorkeeper handling these different scenarios. But the symptoms are exactly the same in both cases. So this leads me to believe it's a problem with our own I18N integration/config.

Summoning @jonatanklosko :mage_man: Only a minor issue, but can be very annoying in UX especially for younger users who see the 2FA screen in Mandarin/Protuguese/Indonesian completely randomly. Do you see what could be going on here at first (or second) glance?

**Comment by jonatanklosko** [comment]
We set the locale in `ApplicationController`

https://github.com/thewca/worldcubeassociation.org/blob/fdec6a0a400bb2a6def1a07eebbaaffe7c307dee/WcaOnRails/app/controllers/application_controller.rb#L26

However, doorkeeper doesn't use this class as it's base controller. Instead it uses so called metal controller (`ActionController::Metal` by default), which is a slimmer controller for improved performance. Consequently, requests handled by the doorkeeper controller don't set the local and instead use whatever `I18n.locale` was last set to in the last regular request handled by the particular application process (so the locale depends on whatever locale people browse the website with, that's why it seems random).

Funny enough, it lead me to https://github.com/doorkeeper-gem/doorkeeper/issues/994, which points back to our older issue https://github.com/thewca/worldcubeassociation.org/issues/1939 :)

**Comment by gregorbg** [comment]
Cool, thanks for the quick feedback! This explains Doorkeeper and the OAuth API, but it doesn't explain Devise and the 2FA issue.

From Jeremy's comment over on the doorkeeper issue that you mentioned, it seems like Devise _should_ be able to read our I18N config correctly. However, I can confidently say that over time, my 2FA prompt has already greeted me in every language currently supported by the WCA website. How do these two issues relate? :thinking:

**Comment by gregorbg** [comment]
For the record, I have just now logged out and back in again to the WCA website three times. Here's the result documented in fresh, out-of-the-oven screenshots:
# Attempt 1: Simplified Chinese
![image](https://user-images.githubusercontent.com/6136469/226020075-13cbd1d5-be99-43dd-8201-0004c2a531e2.png)

# Attempt 2: English (yay, lucky draw!)
![image](https://user-images.githubusercontent.com/6136469/226020411-74b93b8d-17c3-40f8-a4b3-d7641ad66df0.png)

# Attempt 3: Spanish
![image](https://user-images.githubusercontent.com/6136469/226020742-9cadbb81-b43f-4eae-a30c-a251f3ab3513.png)

**Comment by jonatanklosko** [comment]
@gregorbg ah sorry I missed that 2FA is a separate issue. I bet would be that it's because we use `prepend_before_action` and render the page in that action

https://github.com/thewca/worldcubeassociation.org/blob/fdec6a0a400bb2a6def1a07eebbaaffe7c307dee/WcaOnRails/app/controllers/sessions_controller.rb#L4-L5

so this happens before `set_locale` is invoked.

**Comment by dunkOnIT** [comment]
I wasn't able to fully reproduce the 2FA issue on my local machine. Methodology:
* Log in with 2 accounts in 2 browser instances (one normal Firefox, one incognito Firefox), enable 2FA for them
* Set one's language to non-English using language selector in the top-right
* Non-English account: Logout and go through login flow - got English 2FA page, expected non-English 2FA page
* English account: Logout and go through login flow - got English 2FA page, expected non-English 2FA page.

I'm getting some unexpected behaviour, but it basically boils down to the 2FA page never not being English - is there something I would need to do besides changing the language with the language selector to change locale for the non-English account?

**Comment by jonatanklosko** [comment]
@dunkOnIT I would expect the following to reproduce it:

* enter login credentials to get to the 2FA screen
* in another tab visit the homepage and change the locale to something else
* refresh the 2FA screen (possibly a number of times)

**Comment by dunkOnIT** [comment]
@jonatanklosko Using your steps, I wasn't able to reproduce the behaviour if I was in a separate browser/incognito window, but it did work when changing language in a tab in the same browser - but I assume this is because multiple tabs in the same browser are viewed as the same session?

**Comment by jonatanklosko** [comment]
Yeah, I meant an isolated tab like incognito to emulate another client. If those steps are not enough, there may be another factor, but I still think it's likely because of `prepend_before_action`.

**Comment by dunkOnIT** [comment]
I was able to repro this using the following steps: 
- Go to login page and fill in login details for a 2fa-enabled account but do not submit
- With another account set to a different locale [this one was logged in with 2fa, but I don't think that was needed], navigate to any page
- Submit the login details for the 1st account

I've also raised #9810 to fix this issue

**Comment by gregorbg** [comment]
The 2FA issue is being fixed in https://github.com/thewca/worldcubeassociation.org/pull/11175.
I don't think this will fix the OAuth error message though. That one requires a separate PR with a separate solution method

**Comment by gregorbg** [comment]
Fixed by #11176
